specs/interop/superchain_backend.md

```diff
@@ -0,0 +1,156 @@
+block builders and verifiers must locally do so based on their configured level of safety.
+
+## Message Safety Resolution
+
+### Unsafe to Cross-Unsafe
+
+In order to discern between unsafe and cross-unsafe blocks, the backend maintains a graph of
+dependencies between blocks can include interop messages between them.
+
+We describe a psuedo-implementation of cross-unsafe message resolution using this graph. Every node in
+this graph represents a block with edges representing the dependent relationship between them. Upon
+receipt of unsafe payloads, the graph is extended with edges added on detection of interop transactions.
+
+```python
+## Graph Nodes
+blocks = dict() # (hash) -> block
+number_to_block_hash = dict() # (chain_id, int) -> hash
+latest_blocks = dict() # (chain_id) -> hash
+
+## Graph Edge
+##   - The initiating side is represented with the block number since the corresponding payload may
+##   not have been seen at the time the executing message has been processed.
+dependents = dict() # (initiating_chain_id, block_number) -> List[(executing_chain_id, block_hash)]
+dependencies = dict() # (executing_chain_id, block_hash) -> List[(initiating_chain_id, block_number)]
+
+class Block:
+    chain_id: int
+    number: int
+    parent: Block
+
+    executing_messages: List[(MsgId, MsgPayload)]
+    unverified_msg_ref_count: int
+
+def block_safety(chain_id, block_hash):
+    block = blocks[block_hash]
+    if block.unverified_msg_ref_count > 0:
+        return UNSAFE
+
+    ## All messages were validated but we also need to ensure all
+    ## transitive block dependencies have had their messages also
+    ## validated
+    for initiating_chain_id, block_number in dependencies[(chain_id, block_hash)]:
+        if block_bumber > latest_heights[tx.msg_origin]:
+            return UNSAFE
+
+        dependency_block = blocks[number_to_block_hash[(initiating_chain_id, block_number)]]
+        if block_safety(initiating_chain_id, dependency_block.hash) == UNSAFE:
+            return UNSAFE
+
+    return CROSS_UNSAFE
+
+def add_unsafe_payload(chain_id, payload):
+    # **Out of Scope** unsafe reorg handling
+    require(payload.number == blocks[latest_blocks[chain_id]].number + 1)
+
+    blocks[payload.hash] = Block(payload)
+    latest_blocks[chain_id] = payload.hash
+
+    # add edges
+    for tx in blocks[payload.hash].executing_txs:
+        dependents[(tx.msg_origin, tx.block_number)].add((chain_id, payload.hash))
+        dependencies[(chain_id, payload.hash)].add((tx.msg_origin, tx.block_number))
+
+    # try resolve any messages for this block
+    resolve_unverified_messages(chain_id, payload.hash);
+
+    # for existing dependents, trigger resolution
+    for executing_chain_id, block_hash in dependents[(chain_id, payload.number)]:
+        resolve_unverified_messages(executing_chain_id, payload.hash);
+
+def resolve_unverified_messages(chain_id, block_hash):
+    block = blocks[block_hash]
+    for tx in block.executing_txs:
```

Comment:
There is no `executing_txs` property on the block type that you defined. Do you mean `executing_messages`?